### PR TITLE
Fetch a new request token as soon as the browser becomes online

### LINKED ISSUE #16809: CSRF Check failed after being offline for a while
Hi There, 

First up: Dont want to mix things up, so just a remark that it could be related to https://github.com/nextcloud/server/issues/16698 and https://github.com/nextcloud/server/issues/13893

Using an Nextcloud WebApp (like QuickNotes) via Android Firefox, and after being offline for ~120minutes, and then returning i cannot get AJAX requests through, getting CSRF errors.

This is happening with at least 2 apps i tested, so dont assume its an app specific behaviour..

What other information (besides things below) do you need to help pinning this down?

Server Log shows this:
```
{"reqId":"tPO6MMrAB1S5Dcv8xFTW","level":0,"time":"2019-08-20T11:54:17+00:00","remoteAddr":"XYZ.3","user":"XYZ","app":"core","method":"PUT","url":"/nextcloud/index.php/apps/quicknotes/notes/3","message":{"Exception":"OC\\AppFramework\\Middleware\\Security\\Exceptions\\CrossSiteRequestForgeryException","Message":"CSRF check failed","Code":412,"Trace":[{"file":"/var/www/nextcloud/lib/private/AppFramework/Middleware/MiddlewareDispatcher.php","line":95,"function":"beforeController","class":"OC\\AppFramework\\Middleware\\Security\\SecurityMiddleware","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":98,"function":"beforeController","class":"OC\\AppFramework\\Middleware\\MiddlewareDispatcher","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":118,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/Routing/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\QuickNotes\\Controller\\NoteController","update",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"id":"3","_route":"quicknotes.note.update"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"id":"3","_route":"quicknotes.note.update"}]},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"id":"3","_route":"quicknotes.note.update"}]},{"file":"/var/www/nextcloud/lib/base.php","line":987,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/quicknotes/notes/3"]},{"file":"/var/www/nextcloud/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/nextcloud/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php","Line":174,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Android 9; Mobile; rv:68.0) Gecko/68.0 Firefox/68.0","version":"15.0.8.1","id":"5d5bdf6a75e61"}
```

Update:
Also have this error on Chrome for Android:
```
{"reqId":"bX2X65QOf60C7euj8xOP","level":0,"time":"2019-08-20T15:11:11+00:00","remoteAddr":"XYZ.3","user":"XYZ","app":"core","method":"PUT","url":"/nextcloud/index.php/apps/quicknotes/notes/3","message":{"Exception":"OC\\AppFramework\\Middleware\\Security\\Exceptions\\CrossSiteRequestForgeryException","Message":"CSRF check failed","Code":412,"Trace":[{"file":"/var/www/nextcloud/lib/private/AppFramework/Middleware/MiddlewareDispatcher.php","line":95,"function":"beforeController","class":"OC\\AppFramework\\Middleware\\Security\\SecurityMiddleware","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":98,"function":"beforeController","class":"OC\\AppFramework\\Middleware\\MiddlewareDispatcher","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":118,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\QuickNotes\\Controller\\NoteController"},"update"]},{"file":"/var/www/nextcloud/lib/private/AppFramework/Routing/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\QuickNotes\\Controller\\NoteController","update",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"id":"3","_route":"quicknotes.note.update"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"id":"3","_route":"quicknotes.note.update"}]},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"id":"3","_route":"quicknotes.note.update"}]},{"file":"/var/www/nextcloud/lib/base.php","line":987,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/quicknotes/notes/3"]},{"file":"/var/www/nextcloud/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/nextcloud/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php","Line":174,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Linux; Android 9; SM-G950F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.111 Mobile Safari/537.36","version":"15.0.8.1","id":"5d5c0d8f65256"}
```

### Steps to reproduce
1. Show a List of QuickNotes in Webbrowser
2. Let the browser open
3. Set airplaine mode for ~120 minutes
4. Come back online, open the browser again
5. Click on a Note, edit it and try to "save"

### Expected behaviour
Save the Note just fine..

### Actual behaviour
CSRF Error.. 

### Client
- Android 9 Firefox Webbrowser
- Android 9 Chrome Webbrowser

### Server configuration
**Nextcloud version:** (see Nextcloud admin page)
- Instance 1: 15.0.8 (http port 80)
- Instance 2: 16.0.4 (https port 443)

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/server/issues/16809

Debugging a few Sentry tickets I saw evidence of HTTP 412's -> CSRF check failing. Usually those are unexpected as we periodically fetch the current token, but if the client machine is paused or the browser goes offline, the CSRF token might not be fetched when the connection returns. Thus I added event handlers that can pause the background requests when there is no network connectivity and send off a request for the new token as soon as there is connectivity.

Bildschirmfoto von 2020-01-09 14-14-39

How to test
* Open Nextcloud
* Switch browser to offline mode (File -> Work offline on FF, dev tools for Chrome)
* See detection of offline mode
* Restart web server (to force new token)
* Disable offline mode
* See detection and new token in console

As this is something apps might have interest in, e.g. when to resume operation after offline as you have a dependency on the CSRF token, I added events emitted to the event bus where apps can listen to both the browser going offline as well as coming back online.

Do we want this in 18 already? Then we can backport.

### CODE REVIEW

**Comment by rullzer** [comment]
O this is very neat. And also very useful for apps indeed!

I guess 19 is good enough for this for now.

**Comment by tcitworld** [comment]
What about listening to `navigator.onLine` events instead of polling ?

**Comment by kesselb** [comment]
Conflicts :fearful:

**Comment by ChristophWurst** [comment]
> What about listening to `navigator.onLine` events instead of polling ?

We're not polling. We use the online event already to fetch a token when the network state changes. But we need to pool while online as request tokens time out after a certain time.

File: core/src/session-heartbeat.js [hunk 1]
```diff
@@ -63,12 +85,34 @@ export const initSessionHeartBeat = () => {
...
-				console.error('session heartbeat failed', e)
+	window.addEventListener('online', async() => {
+		console.info('browser is online again, resuming heartbeat')
+		interval = startPolling()
+		try {
+			await poll()
+			console.info('session token successfully updated after resuming network')
+
+			// Let apps know we're online and requests will have the new token
+			emit('networkOnline', {
```
Comment (juliushaertl) [inline core/src/session-heartbeat.js]: Would be nice to document those somewhere

**Review by juliusknorr — APPROVED** [review]
Nice :+1:

File: core/src/session-heartbeat.js (hunk 1 above)
Comment (ChristophWurst) [reply core/src/session-heartbeat.js]: Absolutely!

**Comment by rullzer** [comment]
@ChristophWurst can you rebase this? Then we can get this is :)

**Review by gary-kim — APPROVED** [review]
Didn't know about the `online` event. Cool!

**Comment by juliushaertl** [comment]
needs another rebase :dancer:

**Comment by ChristophWurst** [comment]
Now be quick and merge this :P

**Comment by ChristophWurst** [comment]
Rebase cool, rebase fine
Everybody gets a chance
Clap your hands it's party time
Do the rebase dance

**Comment by rullzer** [comment]
jsunit is unhappy... https://drone.nextcloud.com/nextcloud/server/25848/1/2

**Pull request merged**